File: examples/go/inventory/logic/receive_stock.go
package logic

import (
	"angzarr"
	angzarrpb "angzarr/proto/angzarr"
	"inventory/proto/examples"

	goproto "google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/timestamppb"
)

// HandleReceiveStock validates and creates a StockReceived event.
func HandleReceiveStock(cb *angzarrpb.CommandBook, data []byte, state *InventoryState, seq uint32) (*angzarrpb.EventBook, error) {
	var cmd examples.ReceiveStock
	if err := goproto.Unmarshal(data, &cmd); err != nil {
		return nil, angzarr.NewInvalidArgument("failed to unmarshal command: " + err.Error())
	}

	if !state.Exists() {
		return nil, angzarr.NewFailedPrecondition(ErrMsgNotInitialized)
	}
	if cmd.Quantity <= 0 {
		return nil, angzarr.NewInvalidArgument(ErrMsgQuantityPositive)
	}

	return angzarr.PackEvent(cb.Cover, &examples.StockReceived{
		Quantity:   cmd.Quantity,
		NewOnHand:  state.OnHand + cmd.Quantity,
		Reference:  cmd.Reference,
		ReceivedAt: timestamppb.Now(),
	}, seq)
}

File: examples/python/agg-inventory/handlers/receive_stock.py
"""Handler for ReceiveStock command."""

from datetime import datetime, timezone

from google.protobuf.any_pb2 import Any
from google.protobuf.timestamp_pb2 import Timestamp

from angzarr import types_pb2 as types
from errors import CommandRejectedError
from proto import inventory_pb2 as inventory

from handlers.state import InventoryState


def handle_receive_stock(
    command_book: types.CommandBook,
    command_any: Any,
    state: InventoryState,
    seq: int,
) -> types.EventBook:
    """Handle ReceiveStock command."""
    if not state.exists():
        raise CommandRejectedError("Inventory not initialized")

    cmd = inventory.ReceiveStock()
    command_any.Unpack(cmd)

    if cmd.quantity <= 0:
        raise CommandRejectedError("Quantity must be positive")

    event = inventory.StockReceived(
        quantity=cmd.quantity,
        new_on_hand=state.on_hand + cmd.quantity,
        reference=cmd.reference,
        received_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return types.EventBook(
        cover=command_book.cover,
        pages=[
            types.EventPage(
                num=seq,
                event=event_any,
                created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
            )
        ],
    )
